### PR TITLE
process categorical coords properly for multi-glyphs

### LINKED ISSUE #7462: multi_line does not work with categorical plots
#### ALL software version info (bokeh, python, notebook, OS, browser, any other relevant packages)

Testing with Python 2.7.14 and bokeh==0.12.13.

#### Description of expected behavior and the observed behavior

`multi_line` stopped working with categorical plots.

#### Complete, minimal, self-contained example code that reproduces the issue

```python
    x_range = ['a', 'b', 'c', 'd']
    y_values = [1., 2., 3., 4.]
    y_errors = [.1, .2, .3, .4]


    plot = figure(x_range=x_range)
    plot.square(x_range, y_values, size=7, line_alpha=0)

    err_xs = []
    err_ys = []

    for x, y, yerr in zip(x_range, y_values, y_errors):
        err_xs.append((x, x))
        err_ys.append((y - yerr, y + yerr))

    # Option A: Does not work
    # plot.multi_line(err_xs, err_ys)

    # Option B: works!
    for i in xrange(len(err_xs)):
        plot.line(err_xs[i], err_ys[i])
```

Full example in the example repo: https://github.com/sashabaranov/bokeh-leak-example/tree/0.12.13-categorical-multiline

#### Stack traceback and/or browser JavaScript console output

```
[Error] TypeError: undefined is not an object (evaluating 'this._mapping[t[0]].mapping[t[1]]')
	_lookup (bokeh-0.12.13.min.js:1:364819)
	synthetic (bokeh-0.12.13.min.js:1:363563)
	(anonymous function) (bokeh-0.12.13.min.js:1:363701)
	v_synthetic (bokeh-0.12.13.min.js:1:363720)
	set_data (bokeh-0.12.13.min.js:1:235081)
	set_data (bokeh-0.12.13.min.js:1:370298)
	initialize (bokeh-0.12.13.min.js:1:368764)
	e (bokeh-0.12.13.min.js:1:89638)
	e (bokeh-0.12.13.min.js:1:16917)
	e (bokeh-0.12.13.min.js:1:379500)
	e (bokeh-0.12.13.min.js:1:367663)
	build_views (bokeh-0.12.13.min.js:1:14204)
	build_levels (bokeh-0.12.13.min.js:1:345564)
	initialize (bokeh-0.12.13.min.js:1:338801)
	e (bokeh-0.12.13.min.js:1:89638)
	e (bokeh-0.12.13.min.js:1:16917)
	e (bokeh-0.12.13.min.js:1:337113)
	build_views (bokeh-0.12.13.min.js:1:14204)
	build_child_views (bokeh-0.12.13.min.js:1:300626)
	initialize (bokeh-0.12.13.min.js:1:297635)
	e (bokeh-0.12.13.min.js:1:89638)
	e (bokeh-0.12.13.min.js:1:16917)
	e (bokeh-0.12.13.min.js:1:297426)
	e (bokeh-0.12.13.min.js:1:329848)
	i (bokeh-0.12.13.min.js:1:110551)
	add_model_standalone (bokeh-0.12.13.min.js:1:111137)
	r (bokeh-0.12.13.min.js:1:114436)
	(anonymous function) (bokeh-0.12.13.min.js:1:113719)
```

### PR DESCRIPTION
**Pull request by bryevdv**
- [x] issues: fixes #7462
- [x] tests added / passed

### CODE REVIEW

**Comment by bryevdv** [comment]
Integration test:

screen shot 2018-04-02 at 13 37 47

File: bokehjs/src/coffee/models/glyphs/glyph.ts [hunk 1]
```diff
@@ -236,10 +236,33 @@ export abstract class GlyphView extends View {
...
+            const result = []
+            for (const arr of self[xname])
+              result.push(xr.v_synthetic(arr))
+            self[xname] = result
+          }
+          if (yr instanceof FactorRange) {
+            const result = []
+            for (const arr of self[yname])
+              result.push(yr.v_synthetic(arr))
+            self[yname] = result
```
Comment (mattpap) [inline bokehjs/src/coffee/models/glyphs/glyph.ts]: Simpler and faster:

`import {map} from "core/util/arrayable"`

`self[yname] = map(self[yname], (arr) => yr.v_synthetic(arr))`

**Comment by bryevdv** [comment]
Tests are failing with this:
```
Warning: Duplicate __view_model__ or __subtype__ declaration of 'Foo1' for class Foo1.  Previous definition: <class 'bokeh.tests.test_model.test_Model_pretty.<locals>.Foo1'>
```

Which can't possibly be related to these changes, perhaps an upstream pytest change keeping multiple tests in one process?

File: bokehjs/src/coffee/models/glyphs/glyph.ts (hunk 1 above)
Comment (bryevdv) [reply bokehjs/src/coffee/models/glyphs/glyph.ts]: Thanks I will change

Comment (bryevdv) [reply bokehjs/src/coffee/models/glyphs/glyph.ts]: get an error:

> Argument of type '{}' is not assignable to parameter of type 'Arrayable<string | number | [string, string] | [string, string, string] | [string, number] | [string, string, number] | [string, string, string, number]>'. Property 'length' is missing in type '{}'.'

**Comment by mattpap** [comment]
This highlights a problem that bokeh models are identified by name and not module and name pairs.

**Comment by bryevdv** [comment]
> This highlights a problem that bokeh models are identified by name and not module and name pairs.

I think we are free to fix this, I would never agree that `__view_model__` names are public API

**Comment by mattpap** [comment]
So far this wasn't an issue, but we already have abstract models and models sharing a name (abstract Box glyph and Box layout), so at some point we will have a conflict. Obviously that's a bigger change for later time.

**Comment by bryevdv** [comment]
tests pass locally, I am not sure what's going on on Travis

File: bokehjs/src/coffee/models/glyphs/glyph.ts (hunk 1 above)
Comment (bryevdv) [reply bokehjs/src/coffee/models/glyphs/glyph.ts]: This works:
```
if (self._xs != null) {
          if (xr instanceof FactorRange) {
            self[xname] = map(self[xname], (arr: any) => xr.v_synthetic(arr))
          }
          if (yr instanceof FactorRange) {
            self[yname] = map(self[yname], (arr: any) => yr.v_synthetic(arr))
          }
        }
```

Comment (mattpap) [reply bokehjs/src/coffee/models/glyphs/glyph.ts]: Right, this is the effect of allowing `self: any`. I think, for now, `map(self[yname] as any, ...)` should help overcome this error.

Comment (mattpap) [reply bokehjs/src/coffee/models/glyphs/glyph.ts]: Your version is also fine.

**Comment by bryevdv** [comment]
OK restarting a previously passing test fails now so it's definitely some dependency that has changed. All the pytest related packages look identical tho, so I will have to look more carefully at the entire list a bit later.

**Comment by bryevdv** [comment]
This appears to be the difference in installed packages, does not really shed any light offhand:
```
(base) ❯ diff olds news
4a5
>     backcall:                 0.1.0-py36_0
6c7
<     bokeh:                    0.12.15+7.g33955b9-pyh65b294c_0 local
---
>     bokeh:                    0.12.15+6.g97caed3-pyh0280256_0 local
30c31
<     ipython:                  6.2.1-py36h88c514a_1
---
>     ipython:                  6.3.0-py36_0
64c65
<     numpydoc:                 0.7.0-py36h18f165f_0
---
>     numpydoc:                 0.8.0-py36_0
```

**Comment by bryevdv** [comment]
updating `ipython` locally reproduces the problem. I'll make a separate issue.

**Pull request merged**